Event based XML parser for mvnx files

diegoferigo opened the pull request:
Referencing bullet point 3 of robotology-playground/human-dynamics-estimation#15, this commit places the first milestone toward the possibility to replay Xsens suit' data directly from the `.mvnx` file, as described by robotology-playground/human-dynamics-estimation#48.

I'm still working on the following features:

- [x] Refactoring the `getChild()` method in order to avoid manually creating all the `mvnx*` classes for all the xml tags. I'm open to suggestions on this point.
- [x] Allow to select the fields you want to save. Right now all the elements are parsed and stored in the memory

Let me know if you have some comments on the existing codebase.

cc @francesco-romano @traversaro

diegoferigo committed:
Implemented xmlStreamReader class for handling and validating big XMLs

diegoferigo committed:
Inherited xmlStreamReader for mvnx parsing

diegoferigo committed:
Greatly improved the library abstraction

- TEXT and ELEMENT content use the same class
- No need to manually create a class for every xml element
- Implemented the findChildElement() method

diegoferigo committed:
Included the configuration support of mvnxStreamReader

Now it is possible to select what xml elements are stored

diegoferigo committed:
Included Qt dependency in Travis CI

diegoferigo committed:
Style changes

Switched to 4 spaces indentation

diegoferigo committed:
Changed namespaces structure and fixed asserts

diegoferigo committed:
Ported pointers to smart pointers, introduced new typedefs

diegoferigo committed:
Included example for accessing elements with TEXT content

diegoferigo commented:
With the last commit, the parsing part is almost finished.

For the time being, the content of text elements (such as `pos_s`, `connector1`, `connector2`, `orientation`,...) are returned as `string`s. I thought about formatting them in numeric variables, e.g. `vector<float>`, but I eventually reached the conclusion that this step should be performed by the client. Thinking this as a library, it cannot know what data type the user want to use to manipulate the data (`std::vector`, `Eigen::VectorXd`,...). Comments?

@francesco-romano @claudia-lat @traversaro

traversaro commented on mvnxparser/CMakeLists.txt:
```diff
@@ -0,0 +1,21 @@
+# Author: Diego Ferigo
+# CopyPolicy: Released under the terms of the LGPLv2.1 or later, see LGPL.TXT
+
+set(ENABLE_mvnx_parser ON CACHE BOOL "Enable the Qt-based XML parser")
```
Just a curiosity: why you did not use an option?

diegoferigo replied:
No particular reason, switched to option() in the next commit

traversaro commented on mvnxparser/mvnxStreamReader/CMakeLists.txt:
```diff
@@ -0,0 +1,16 @@
+# Load dependencies
+find_package(Qt5Xml REQUIRED)
+
+# Build the libraries
```
Any reason for hard coding the use of a STATIC library?

diegoferigo replied:
I hardcoded it in the first tests I did, just removed it in all the files of this folder tree

traversaro replied:
The library type is managed by `BUILD_SHARED_LIBS` variable, that defaults to ON, hence SHARED. 
To avoid problems in compiling the library as shared in Windows, I suggest moving the calls setting `CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS` (https://github.com/robotology-playground/xsens-mvn/search?utf8=%E2%9C%93&q=CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS&type=) to the root CMakeLists.txt to make sure it is always visible.

traversaro commented on mvnxparser/mvnxStreamReader/CMakeLists.txt:
```diff
@@ -0,0 +1,16 @@
+qt5_use_modules(mvnxStreamReader Xml)
+
+# Export the includes needed to who inherits this library
+target_include_directories(mvnxStreamReader INTERFACE ${CMAKE_CURRENT_SOURCE_DIR})
```
Consider that the INTERFACE include directories will be propagated to all the target linking to this target, even when it is installed. I suggest to guard the use of this local include directory with a BUILD specific generator expression.

diegoferigo replied:
The install target has not yet been implemented. I'll address this comment as soon as the install support will be included.

traversaro reviewed (approved):
I have some questions but they are not blocking.

francesco-romano commented on mvnxparser/CMakeLists.txt:
```diff
@@ -0,0 +1,21 @@
+
+    # Build the executables
+    # XML
+    add_executable(xmlStreamReaderDriver xmlStreamReaderDriver.cpp)
```
If I understood correctly you implemented two libraries and two executables.
As probably we will use the libraries inside other modules or YARP devices, do you mind splitting the libraries and executable into two different folders? 
Or, if the executables are "test" executables, then probably a subfolder inside this folder is sufficient.

diegoferigo replied:
Moved them in Test subfolders.

francesco-romano commented on mvnxparser/xmlStreamReaderDriver.cpp:
```diff
@@ -0,0 +1,49 @@
+int main(int argc, char* argv[])
+{
+    xmlStreamReader parser;
+    if (parser.setDocument("/home/dferigo/git/xml-demo-qt/Meri-020.mvnx")) {
```
hardcoded path

francesco-romano commented on mvnxparser/xmlStreamReaderDriver.cpp:
```diff
@@ -0,0 +1,49 @@
+        return 1;
+    }
+
+    if (parser.setSchema("/home/dferigo/git/xml-demo-qt/schema.xsd"))
```
hardcoded path

francesco-romano commented on mvnxparser/xmlStreamReader/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+find_package(Qt5XmlPatterns REQUIRED)
+
+# Build libraries
+add_subdirectory(xmlMessageHandler)
```
Do you need to have a single library of this? Why don't put the sources inside the `xmlStreamReader` library?

diegoferigo replied:
The message handler is the responsible of formatting how errors are displayed. It is not mandatory and can be customized by the user. I implemented a very basic class, and it made sense adding it inside an external folder for this reason. I don't mind shipping a _default_ error message, and then let the user know that the `xmlMessageHandler` class can be inherited and the `printMessage()` method overloaded.

diegoferigo committed:
Minor changes in CMakeLists

Removed the hardcoded STATIC option for libraries
Enable / disable this library using option()
Export symbols in Windows builds

francesco-romano commented on mvnxparser/mvnxStreamReader/mvnxStreamReader.h:
```diff
@@ -0,0 +1,70 @@
+// Simple container of the mvnx parsing configuration.
+typedef std::unordered_map<std::string, bool> mvnxConf;
+
+class mvnxStreamReader : public xmlstream::xmlStreamReader {
```
I like classes to start capital case.
Given that MVNX is an acronym, I would call  the class
`MVNXStreamReader `.

This also for the other classes

francesco-romano commented on mvnxparser/mvnxStreamReader/mvnxStreamReader.h:
```diff
@@ -0,0 +1,70 @@
+namespace mvnx {
+
+// Simple container of the mvnx parsing configuration.
+typedef std::unordered_map<std::string, bool> mvnxConf;
```
See comment below. I think that applies also to `typedef` as you are defining an alias to a type.
I would be also more verbose, i.e. `MVNXConfiguration`

francesco-romano commented on mvnxparser/mvnxStreamReader/mvnxStreamReader.h:
```diff
@@ -0,0 +1,70 @@
+    xmlContentPtrS getXmlTreeRoot() const
+    {
+        return std::dynamic_pointer_cast<xmlContent>(xmlTreeRoot);
+    };
```
semicolon after method definition is not necessary.
(this comment added only once).
If you don't mind, please move the definition in the associated `.cpp` file

francesco-romano commented on mvnxparser/mvnxStreamReader/mvnxStreamReader.cpp:
```diff
@@ -0,0 +1,218 @@
+mvnxStreamReader::processAttributes(QXmlStreamAttributes elementAttributes)
+{
+    attributes_t attributes;
+    if (not elementAttributes.empty()) {
```
Is there a reason why you used `not` and not `!`?
If you don't mind I prefer the `!`

diegoferigo replied:
I like `not`, it is more human readable, and sources are made for humans beings :) I don't mind using `!` tho.

francesco-romano commented on mvnxparser/mvnxStreamReader/mvnxStreamReader.cpp:
```diff
@@ -0,0 +1,218 @@
+{
+    attributes_t attributes;
+    if (not elementAttributes.empty()) {
+        for (auto element : elementAttributes) {
```
You can also use `const auto& element` so using a reference and avoiding a copy

diegoferigo replied:
`const` is for sure safer, I agree

francesco-romano commented on mvnxparser/mvnxStreamReader/mvnxStreamReader.cpp:
```diff
@@ -0,0 +1,218 @@
+void mvnxStreamReader::handleCharacters(string elementText)
+{
+    IContentPtrS lastElement = elementsLIFO.back();
+    assert(lastElement != nullptr);
```
I prefer checks to `nullptr` or `0` to be implicit, i.e. `assert(lastElement)`

diegoferigo replied:
I agree, this is a leftover from the old version without smart pointers

francesco-romano commented on mvnxparser/xmlStreamReader/xmlStreamReader.cpp:
```diff
@@ -0,0 +1,67 @@
+
+xmlStreamReader::xmlStreamReader(string documentFile, string schemaFile)
+    : xmlStreamReader(documentFile)
+{
```
Just a curiosity.
Usually the designated initialiser is the more complete one, not the contrary.
Why this choice here?

diegoferigo replied:
Different way of doing + `C++11` features. What I like of the contrary is the single location of the code, what I dislike is the need of initializing the less complete constructor with dummy arguments. Considering this small trade-off, for the sake of maintainability of the code, I think that your suggestion has a slight advantage, empty-body constructors are more elegant.

francesco-romano commented on mvnxparser/mvnxStreamReader/mvnxStreamReader.cpp:
```diff
@@ -0,0 +1,218 @@
+        elementAttributes = xml.attributes();
+
+        // Handle the generated events
+        switch (tokenType) {
```
Given the limited options we need for parsing MVNX files this is ok.
Probably if we have to code a more generic XML parsing library I would try to opt to a visitor pattern

diegoferigo replied:
This code section is QT-related, and those are the events that their implementation of the SAX parser (even if QT implements something not strictly SAX, but you can think it as SAX) is throwing. This library is based on those events, and it builds all the logic on top of them.

francesco-romano reviewed (commented):
See inline comments

diegoferigo committed:
Style Edits

- Changed files and folders names
- Changed classes names
- Minor edits in constructors, and const variables

diegoferigo committed:
Moved test executables

diegoferigo reviewed (commented):
I have addressed your comments with the last commits.

francesco-romano commented on mvnxparser/MVNXStreamReader/CMakeLists.txt:
```diff
@@ -0,0 +1,21 @@
+find_package(Qt5Xml REQUIRED)
+
+# Build the libraries
+add_library(MVNXStreamReader MVNXStreamReader.cpp)
```
Class names should be CamelCase (capitalized).

But for library names... I don't think that applies to. @traversaro what do you think?

traversaro replied:
I never saw any consistency in library names.. 
just to see some example from my codyco-superbuild : 
`YARP_OS`, `gazebo_yarp_camera`, `idyntree-core`, `idyntree-core`, `skinDynLib`, `wholeBodyDynamics_IDLServer`. Several variability can be found on system libraries. I am agnostic to the style used in the library names.

traversaro commented:
Related to https://github.com/robotology-playground/xsens-mvn/pull/12#issuecomment-294855499 @diegoferigo returning strings for content of the elements I think it is the right thing to do unless we have a concrete use case for which it is convenient otherwise.

traversaro commented:
Related to https://github.com/robotology-playground/xsens-mvn/pull/12#discussion_r112155821 .
Regarding the `not` operator, I also like it visually. However consider that even if it is in the standard, to use it in Visual C++ you need to either include the `<ciso646>` header or pass the `/Za` flag to the compiler [1]. In general disabling compiler-specific extensions  is a good idea, so adding `/Za` consistently to our projects could make sense, but this needs to be done and tested. : ) 

[1] : https://msdn.microsoft.com/en-us/library/1k6w8551.aspx

diegoferigo committed:
Added a weak_ptr to parent element

diegoferigo committed:
Added support for install target

francesco-romano commented:
@diegoferigo can you modify the CI configurations files also of the projects that depend on this? (I am referring to https://github.com/robotology-playground/human-dynamics-estimation, see https://travis-ci.org/robotology-playground/human-dynamics-estimation/jobs/228010561 is failing now
